From 86ed61eb1334f624e380467c081b02ac18b6805d Mon Sep 17 00:00:00 2001
From: Uma Shankar <uma.shankar@intel.com>
Date: Thu, 8 Jan 2015 15:36:14 +0530
Subject: [PATCH 1/1] FOR_UPSTREAM [VPG]: drm/i915: Rectified generic PMIC
 parsing logic

Generic PMIC parsing logic was not compatible with the
latest VBT interface document. Rectified the same.

Issue: IRDA-2822
Change-Id: Ib1738984ea4596a2714162523fcf9c02b4f47944
Signed-off-by: Uma Shankar <uma.shankar@intel.com>
Signed-off-by: Li, Alex C <alex.c.li@intel.com>
---
 drivers/gpu/drm/i915/intel_dsi_panel_vbt.c | 23 ++++++++++++++++++-----
 1 file changed, 18 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_dsi_panel_vbt.c b/drivers/gpu/drm/i915/intel_dsi_panel_vbt.c
index c78ac58..74a058d 100644
--- a/drivers/gpu/drm/i915/intel_dsi_panel_vbt.c
+++ b/drivers/gpu/drm/i915/intel_dsi_panel_vbt.c
@@ -440,12 +440,16 @@ static u8 *mipi_exec_spi(struct intel_dsi *intel_dsi, u8 *data)
 
 static u8 *mipi_exec_pmic(struct intel_dsi *intel_dsi, u8 *data)
 {
-	u8 pmic_page;
 	u32 register_address, register_data;
-	u32 data_mask, tmp;
+	int data_mask, tmp;
+	int ret;
 
-	data++;
-	pmic_page = *data++;
+	/*
+	 * First 3 bytes are not relevant for Linux.
+	 * Skipping the data field by 3 bytes to get
+	 * the PMIC register Address.
+	 */
+	data += 3;
 	register_address = *((u32 *)data);
 	data += 4;
 	register_data = *((u32 *)data);
@@ -454,10 +458,19 @@ static u8 *mipi_exec_pmic(struct intel_dsi *intel_dsi, u8 *data)
 	data += 4;
 
 	tmp = intel_soc_pmic_readb(register_address);
+	if (tmp < 0) {
+		DRM_ERROR("PMIC Read failed\n");
+		return ERR_PTR(tmp);
+	}
+
 	tmp &= ~data_mask;
 	register_data &= data_mask;
 	register_data |= tmp;
-	intel_soc_pmic_writeb(register_address, register_data);
+	ret = intel_soc_pmic_writeb(register_address, register_data);
+	if (ret < 0) {
+		DRM_ERROR("PMIC Write failed\n");
+		return ERR_PTR(ret);
+	}
 
 	return data;
 }
-- 
1.9.1

